File: packages/reown_yttrium/ios/Classes/Extensions.swift
import YttriumWrapper

extension Eip1559Estimation {
    func toJson() -> [String: Any] {
        return [
            "maxFeePerGas": self.maxFeePerGas,
            "maxPriorityFeePerGas": self.maxPriorityFeePerGas
        ]
    }
}

extension SolanaTransaction {
    func toJson() -> [String: Any] {
        return [
            "chainId": chainId,
            "from": from,
            "transaction": transaction
        ]
    }
}

extension SolanaTxnDetails {
    func toJson() -> [String: Any] {
        return [
            "transaction": transaction.toJson(),
            "transactionHashToSign": transactionHashToSign
        ]
    }
}

extension Route {
    func toJson() -> [[String: Any]] {
        switch self {
        case .eip155(let txns):
            return (txns as [TxnDetails]).map { $0.toJson() } // list of TxnDetails
        case .solana(let txns):
            return (txns as [SolanaTxnDetails]).map { $0.toJson() } // list of SolanaTxnDetails
        }
    }
}

extension UiFields {
    func toJson() -> [String: Any] {
        let route = route.map { $0.toJson() }.first
        return [
            "routeResponse": routeResponse.toJson(),
            "route": route ?? [],
            "localRouteTotal": localRouteTotal.toJson(),
            "bridge": bridge.map { $0.toJson() },
            "localBridgeTotal": localBridgeTotal.toJson(),
            "initial": initial.toJson(),
            "localTotal": localTotal.toJson()
        ]
    }
}

extension Transactions {
    func toJson() -> [[String: Any]] {
        switch self {
        case .eip155(let txns):
            return (txns as [Transaction]).map { $0.toJson() } // list of Transaction
        case .solana(let txns):
            return (txns as [SolanaTransaction]).map { $0.toJson() } // list of SolanaTransaction
        }
    }
}

extension PrepareResponseAvailable {
    func toJson() -> [String: Any] {
        let transactions = transactions.map { $0.toJson() }.first // list of `Transactions`
        return [
            "orchestrationId": orchestrationId,
            "initialTransaction": initialTransaction.toJson(),
            "transactions": transactions ?? [],
            "metadata": metadata.toJson()
        ]
    }
}

extension Transaction {
    func toJson() -> [String: Any] {
        return [
            "chainId": chainId,
            "from": from,
            "to": to,
            "value": value,
            "input": input,
            "gasLimit": gasLimit,
            "nonce": nonce
        ]
    }
}

extension Metadata {
    func toJson() -> [String: Any] {
        return [
            "fundingFrom": fundingFrom.map { $0.toJson() },
            "initialTransaction": initialTransaction.toJson(),
            "checkIn": String(checkIn)
        ]
    }
}

extension FundingMetadata {
    func toJson() -> [String: Any] {
        return [
            "chainId": chainId,
            "tokenContract": tokenContract,
            "symbol": symbol,
            "amount": amount,
            "bridgingFee": bridgingFee,
            "decimals": decimals
        ]
    }
}

extension InitialTransactionMetadata {
    func toJson() -> [String: Any] {
        return [
            "transferTo": transferTo,
            "amount": amount,
            "tokenContract": tokenContract,
            "symbol": symbol,
            "decimals": decimals
        ]
    }
}

extension TxnDetails {
    func toJson() -> [String: Any] {
        return [
            "transaction": transaction.toJson(),
            "transactionHashToSign": transactionHashToSign,
            "fee": fee.toJson()
        ]
    }
}

extension FeeEstimatedTransaction {
    func toJson() -> [String: Any] {
        return [
            "chainId": chainId,
            "from": from,
            "to": to,
            "value": value,
            "input": input,
            "gasLimit": gasLimit,
            "nonce": nonce,
            "maxFeePerGas": maxFeePerGas,
            "maxPriorityFeePerGas": maxPriorityFeePerGas
        ]
    }
}

extension TransactionFee {
    func toJson() -> [String: Any] {
        return [
            "fee": fee.toJson(),
            "localFee": localFee.toJson()
        ]
    }
}

extension PrepareResponseNotRequired {
    func toJson() -> [String: Any] {
        return [
            "initialTransaction": initialTransaction.toJson(),
            "transactions": transactions.map { $0.toJson() }
        ]
    }
}

extension PrepareResponseError {
    func toJson() -> [String: Any] {
        return [
            "error": error.toJson(), // BridgingError
            "reason": reason,
        ]
    }
}

extension BridgingError {
    func toJson() -> String {
        switch self {
        case .assetNotSupported:
            return "assetNotSupported"
        case .noRoutesAvailable:
            return "noRoutesAvailable"
        case .insufficientFunds:
            return "insufficientFunds"
        case .insufficientGasFunds:
            return "insufficientGasFunds"
        default:
            return "unknown"
        }
    }
}

extension Amount {
    func toJson() -> [String: Any] {
        return [
            "symbol": symbol,
            "amount": amount,
            "unit": unit,
            "formatted": formatted,
            "formattedAlt": formattedAlt
        ]
    }
}

extension ExecuteDetails {
    func toJson() -> [String: Any] {
        return [
            "initialTxnReceipt": initialTxnReceipt,
            "initialTxnHash": initialTxnHash
        ]
    }
}

extension Currency {
    static func fromString(value: String) -> Currency {
        switch value.lowercased() {
        case "usd": return .usd
        case "eur": return .eur
        case "gbp": return .gbp
        case "aud": return .aud
        case "cad": return .cad
        case "inr": return .inr
        case "jpy": return .jpy
        case "btc": return .btc
        case "eth": return .eth
        default: return .usd
        }
    }
}

File: packages/reown_yttrium_utils/android/src/main/kotlin/com/reown/reown_yttrium_utils/Extensions.kt
package com.reown.reown_yttrium_utils

import kotlinx.serialization.json.*
import uniffi.uniffi_yttrium_utils.Eip1559Estimation
import uniffi.yttrium_utils.Amount
import uniffi.yttrium_utils.BridgingError
import uniffi.yttrium_utils.ExecuteDetails
import uniffi.yttrium_utils.FeeEstimatedTransaction
import uniffi.yttrium_utils.FundingMetadata
import uniffi.yttrium_utils.InitialTransactionMetadata
import uniffi.yttrium_utils.PrepareResponseAvailable
import uniffi.yttrium_utils.PrepareResponseError
import uniffi.yttrium_utils.PrepareResponseMetadata
import uniffi.yttrium_utils.PrepareResponseNotRequired
import uniffi.yttrium_utils.Transaction
import uniffi.yttrium_utils.TransactionFee
import uniffi.yttrium_utils.TxnDetails
import uniffi.yttrium_utils.SolanaTxnDetails
import uniffi.yttrium_utils.UiFields
import uniffi.yttrium_utils.Route
import uniffi.yttrium_utils.SendTxMessage
import uniffi.yttrium_utils.SolanaTransaction
import uniffi.yttrium_utils.Transactions

fun Eip1559Estimation.toMap(): Map<String, Any> {
    return mapOf(
        "maxFeePerGas" to maxFeePerGas,
        "maxPriorityFeePerGas" to maxPriorityFeePerGas
    )
}

fun SolanaTransaction.toMap(): Map<String, Any> {
    return mapOf(
        "chainId" to chainId,
        "from" to from,
        "transaction" to transaction
    )
}

fun SolanaTxnDetails.toMap(): Map<String, Any> {
    return mapOf(
        "transaction" to transaction.toMap(), // SolanaTransaction does not have toMap() so it's done here
        "transactionHashToSign" to transactionHashToSign
    )
}

fun Route.toMap(): List<Map<String, Any>> = when (this) {
    is Route.Eip155 -> v1.map { it.toMap() } // List<TxnDetails>
    is Route.Solana -> v1.map { it.toMap() } // List<SolanaTxnDetails>
}

fun UiFields.toMap(): Map<String, Any> {
    val route = route.map { it.toMap() }.first() // Route does not have toMap() so it's done here
    return mapOf(
        "routeResponse" to routeResponse.toMap(),
        "route" to route,
        "localRouteTotal" to localRouteTotal.toMap(),
        "bridge" to bridge.map { it.toMap() },
        "localBridgeTotal" to localBridgeTotal.toMap(),
        "initial" to initial.toMap(),
        "localTotal" to localTotal.toMap()
    )
}

fun Transactions.toMap(): List<Map<String, Any>> = when (this) {
    is Transactions.Eip155 -> v1.map { it.toMap() } // List<Transaction>
    is Transactions.Solana -> v1.map { it.toMap() } // List<SolanaTransaction>, SolanaTransaction has toMap() defined here
}

fun PrepareResponseAvailable.toMap(): Map<String, Any> {
    val transactions = transactions.map { it.toMap() }.first() // Route does not have toMap() so it's done here
    return mapOf(
        "orchestrationId" to orchestrationId,
        "initialTransaction" to initialTransaction.toMap(),
        "transactions" to transactions,
        "metadata" to metadata.toMap()
    )
}

fun Transaction.toMap(): Map<String, Any> {
    return mapOf(
        "chainId" to chainId,
        "from" to from,
        "to" to to,
        "value" to value,
        "input" to input,
        "gasLimit" to gasLimit,
        "nonce" to nonce
    )
}

fun PrepareResponseMetadata.toMap(): Map<String, Any> {
    return mapOf(
        "fundingFrom" to fundingFrom.map { it.toMap() },
        "initialTransaction" to initialTransaction.toMap(),
        "checkIn" to checkIn.toString()
    )
}

fun FundingMetadata.toMap(): Map<String, Any> {
    return mapOf(
        "chainId" to chainId,
        "tokenContract" to tokenContract,
        "symbol" to symbol,
        "amount" to amount,
        "bridgingFee" to bridgingFee,
        "decimals" to decimals.toInt() // TODO UByte
    )
}

fun InitialTransactionMetadata.toMap(): Map<String, Any> {
    return mapOf(
        "transferTo" to transferTo,
        "amount" to amount,
        "tokenContract" to tokenContract,
        "symbol" to symbol,
        "decimals" to decimals.toInt() // TODO UByte
    )
}

fun TxnDetails.toMap(): Map<String, Any> {
    return mapOf(
        "transaction" to transaction.toMap(),
        "transactionHashToSign" to transactionHashToSign,
        "fee" to fee.toMap()
    )
}

fun FeeEstimatedTransaction.toMap(): Map<String, Any> {
    return mapOf(
        "chainId" to chainId,
        "from" to from,
        "to" to to,
        "value" to value,
        "input" to input,
        "gasLimit" to gasLimit,
        "nonce" to nonce,
        "maxFeePerGas" to maxFeePerGas,
        "maxPriorityFeePerGas" to maxPriorityFeePerGas
    )
}

fun TransactionFee.toMap(): Map<String, Any> {
    return mapOf(
        "fee" to fee.toMap(),
        "localFee" to localFee.toMap()
    )
}

fun PrepareResponseNotRequired.toMap(): Map<String, Any> {
    return mapOf(
        "initialTransaction" to initialTransaction.toMap(),
        "transactions" to transactions.map { it.toMap() }
    )
}

fun PrepareResponseError.toMap(): Map<String, Any> {
    return mapOf(
        "error" to error.toMap(),
        "reason" to reason
    )
}

fun BridgingError.toMap(): String = when (this) {
    BridgingError.ASSET_NOT_SUPPORTED -> "assetNotSupported"
    BridgingError.NO_ROUTES_AVAILABLE -> "noRoutesAvailable"
    BridgingError.INSUFFICIENT_FUNDS -> "insufficientFunds"
    BridgingError.INSUFFICIENT_GAS_FUNDS -> "insufficientGasFunds"
//    BridgingError.TRANSACTION_SIMULATION_FAILED -> "transactionSimulationFailed"
    else -> this.name.toString()
}

fun Amount.toMap(): Map<String, Any> {
    return mapOf(
        "symbol" to symbol,
        "amount" to amount,
        "unit" to unit.toInt(), // TODO UByte
        "formatted" to formatted,
        "formattedAlt" to formattedAlt
    )
}

fun ExecuteDetails.toMap(): Map<String, *> {
    val json = Json.encodeToString(this)
    val elem = Json.parseToJsonElement(json)
    return (elem as JsonObject).mapValues { Json.encodeToString(it.value) }
}

fun Map<String, Any?>.toSendTxMessage(): SendTxMessage {
    val jsonString = mapToJsonString(this)
    return Json.decodeFromString<SendTxMessage>(jsonString)
}

fun List<*>.toSendTxMessageList(): List<SendTxMessage> {
    return this.mapNotNull { item ->
        (item as? Map<String, Any?>)?.toSendTxMessage()
    }
}

/** JSON string -> Map<String, Any?> (single function) */
fun jsonStringToMap(json: String): Map<String, Any?> {
    val root = Json.parseToJsonElement(json)
    require(root is JsonObject) { "Root JSON element must be an object" }
    fun elemToAny(e: JsonElement): Any? = when (e) {
        is JsonNull -> null
        is JsonPrimitive -> e.booleanOrNull ?: e.intOrNull ?: e.longOrNull ?: e.doubleOrNull ?: e.content
        is JsonObject -> e.mapValues { (_, v) -> elemToAny(v) }
        is JsonArray -> e.map { elemToAny(it) }
    }
    return root.mapValues { (_, v) -> elemToAny(v) }
}

/** Map<String, Any?> -> JSON string (single function) */
fun mapToJsonString(map: Map<String, Any?>): String {
    fun anyToElem(value: Any?): JsonElement = when (value) {
        null -> JsonNull
        is Boolean -> JsonPrimitive(value)
        is Number -> JsonPrimitive(value)
        is String -> JsonPrimitive(value)
        is Map<*, *> -> buildJsonObject { value.forEach { (k, v) -> put(k.toString(), anyToElem(v)) } }
        is Collection<*> -> buildJsonArray { value.forEach { add(anyToElem(it)) } }
        is Array<*> -> buildJsonArray { value.forEach { add(anyToElem(it)) } }
        else -> JsonPrimitive(value.toString())
    }
    return Json.encodeToString(anyToElem(map))
}